packages/client/src/base/action-handler-registry.ts

```diff
@@ -14,56 +14,57 @@
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
  ********************************************************************************/
 
-import { ActionHandlerRegistration, ActionHandlerRegistry, IActionHandler, IActionHandlerInitializer, TYPES } from '@eclipse-glsp/sprotty';
-import { inject, injectable, named } from 'inversify';
-import { IContributionProvider } from '../utils/contribution-provider';
+import {
+    ActionHandlerRegistration,
+    ActionHandlerRegistry,
+    BindingContext,
+    IActionHandlerInitializer,
+    LazyInjector,
+    TYPES,
+    bindOrRebind
+} from '@eclipse-glsp/sprotty';
+import { decorate, inject, injectable, unmanaged } from 'inversify';
 
 @injectable()
 export class GLSPActionHandlerRegistry extends ActionHandlerRegistry {
-    @inject(TYPES.IContributionProvider)
-    @named(TYPES.ActionHandlerRegistration)
-    protected readonly registrations: IContributionProvider<ActionHandlerRegistration>;
-
-    @inject(TYPES.IContributionProvider)
-    @named(TYPES.IActionHandlerInitializer)
-    protected readonly initializers: IContributionProvider<IActionHandlerInitializer>;
+    @inject(LazyInjector)
+    protected lazyInjector: LazyInjector;
 
     protected initialized = false;
 
     constructor() {
         super([], []);
     }
-
-    protected init(): void {
-        if (!this.initialized) {
-            this.initialized = true;
-            this.registrations.getContributions().forEach(registration => this.register(registration.actionKind, registration.factory()));
-            this.initializers.getContributions().forEach(initializer => this.initializeActionHandler(initializer));
-        }
-    }
-
-    override register(key: string, instance: IActionHandler): void {
-        this.init();
-        super.register(key, instance);
-    }
-
-    override get(key: string): IActionHandler[] {
-        this.init();
-        return super.get(key);
-    }
-
-    override initializeActionHandler(initializer: IActionHandlerInitializer): void {
-        this.init();
-        super.initializeActionHandler(initializer);
-    }
-
     /**
      * Retrieve a set of all action kinds for which (at least) one
      * handler is registered
      * @returns the set of handled action kinds
      */
     getHandledActionKinds(): string[] {
-        this.init();
         return Array.from(this.elements.keys());
     }
+
+    initialize(): void {
+        if (this.initialized) {
+            return;
+        }
+
+        this.lazyInjector
+            .getAll<ActionHandlerRegistration>(TYPES.ActionHandlerRegistration)
+            .forEach(registration => this.register(registration.actionKind, registration.factory()));
+        this.lazyInjector
+            .getAll<IActionHandlerInitializer>(TYPES.IActionHandlerInitializer)
+            .forEach(initializer => this.initializeActionHandler(initializer));
+    }
+}
+
+let baseClassDecorated = false;
```

Comment:
I've seen this pattern a few times now, is this something that can be added to our DI utilities (e.g., `decorateUnamanged`)? We could keep a map of targets that we already re-decorated with unmanaged. Just because it is basically always the same code with just the target and the parameter indices as arguments.

Reply:
After retesting this, I don't think its necessary at all. Just passing an empty array in the constructor seems to be enough.
Seem like there was another underyling issue when I originally tested this and then introduced the fix. But with the current solution we don't need the post decoration at all.

Reply:
I really like the new solution much better, great idea and work!